File: MythicDrops/src/main/java/net/nunnerycode/bukkit/mythicdrops/repair/MythicRepairCost.java
package net.nunnerycode.bukkit.mythicdrops.repair;

import net.nunnerycode.bukkit.mythicdrops.api.items.MythicItemStack;
import net.nunnerycode.bukkit.mythicdrops.api.repair.RepairCost;
import org.bukkit.inventory.ItemStack;
import org.bukkit.material.MaterialData;

import java.util.List;

public final class MythicRepairCost implements RepairCost {
	private final String name;
	private final int priority;
	private final int experienceCost;
	private final double repairPercentagePerCost;
	private final int amount;
	private final MaterialData materialData;
	private final String itemName;
	private final List<String> itemLore;

	public MythicRepairCost(String name, int priority, int experienceCost, double repairPercentagePerCost, int amount,
							MaterialData materialData, String itemName, List<String> itemLore) {
		this.name = name;
		this.priority = priority;
		this.experienceCost = experienceCost;
		this.repairPercentagePerCost = repairPercentagePerCost;
		this.amount = amount;
		this.materialData = materialData;
		this.itemName = itemName;
		this.itemLore = itemLore;
	}

	@Override
	public List<String> getItemLore() {
		return itemLore;
	}

	@Override
	public String getItemName() {
		return itemName;
	}

	@Override
	public MaterialData getMaterialData() {
		return materialData;
	}

	@Override
	public int getAmount() {
		return amount;
	}

	@Override
	public double getRepairPercentagePerCost() {
		return repairPercentagePerCost;
	}

	@Override
	public int getExperienceCost() {
		return experienceCost;
	}

	@Override
	public int getPriority() {
		return priority;
	}

	@Override
	public String getName() {
		return name;
	}

	@Override
	public ItemStack toItemStack(int amount) {
		return new MythicItemStack(materialData.getItemType(), amount, (short) 0,
				(itemName == null || itemName.isEmpty()) ? null : itemName,
				(itemLore == null || itemLore.isEmpty()) ? null : itemLore);
	}

}

File: mythicdrops/src/main/kotlin/com/tealcube/minecraft/bukkit/mythicdrops/repair/MythicRepairCost.kt
/*
 * This file is part of MythicDrops, licensed under the MIT License.
 *
 * Copyright (C) 2019 Richard Harrah
 *
 * Permission is hereby granted, free of charge,
 * to any person obtaining a copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software,
 * and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS
 * OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package com.tealcube.minecraft.bukkit.mythicdrops.repair

import com.squareup.moshi.JsonClass
import com.tealcube.minecraft.bukkit.mythicdrops.api.repair.RepairCost
import com.tealcube.minecraft.bukkit.mythicdrops.setDisplayNameChatColorized
import com.tealcube.minecraft.bukkit.mythicdrops.setLoreChatColorized
import org.bukkit.Material
import org.bukkit.enchantments.Enchantment
import org.bukkit.inventory.ItemStack

@JsonClass(generateAdapter = true)
data class MythicRepairCost(
    override val itemLore: List<String>?,
    override val itemName: String?,
    override val material: Material,
    override val amount: Int,
    override val repairPercentagePerCost: Double,
    override val experienceCost: Int,
    override val priority: Int,
    override val name: String,
    override val enchantments: Map<Enchantment, Int>?
) : RepairCost {
    override fun toItemStack(amount: Int): ItemStack = ItemStack(material, amount).let {
        if (itemName != null) {
            it.setDisplayNameChatColorized(itemName)
        }
        if (itemLore != null) {
            it.setLoreChatColorized(itemLore)
        }
        it.enchantments.forEach { (enchantment, _) -> it.removeEnchantment(enchantment) }
        if (enchantments != null) {
            it.addUnsafeEnchantments(enchantments)
        }
        it
    }
}
